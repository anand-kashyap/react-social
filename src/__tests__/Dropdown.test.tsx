import React from 'react'
import { render, cleanup, fireEvent } from '@testing-library/react'
import Dropdown, { dopt } from 'utils/dropdown/DropDown';

const renderDropdown = (dropOpts: dopt[], selOpt: any = () => { }, id = 'test1') => {
  return render(
    <Dropdown id={id} dropOpts={dropOpts} selOpt={selOpt} />
  );
};

function addOffsetParent(dBtn: HTMLElement) {
  Object.defineProperty(dBtn, 'offsetParent', {
    writable: true,
    value: {
      getBoundingClientRect() {
        return {
          x: 851.671875,
          y: 200.046875,
          width: 8.34375,
          height: 17,
          top: 967.046875,
          right: 860.015625,
          toJSON() { },
          bottom: 984.046875,
          left: 851.671875,
        };
      }
    },
  });
}
describe('Dropdown Component', () => {
  afterEach(cleanup);
  const defOps: dopt[] = [
    { val: 'one', icon: 'all_out' },
    { val: 'delete', icon: 'delete' }
  ];

  it('should take a snapshot', () => {
    const { asFragment } = renderDropdown(defOps);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should toggle container on button click', () => {
    const { getByTestId } = renderDropdown(defOps),
      dBtn = getByTestId('drop-btn'),
      dCtn = getByTestId('drop-contain');
    addOffsetParent(dBtn);
    expect(dCtn.style.display).toBeFalsy();
    fireEvent.click(dBtn);
    expect(dCtn.style.display).toBeTruthy();
  });

  it('should return correct value on dropdown opt click', () => {
    const mockFn = jest.fn(v => v),
      { getByTestId } = renderDropdown(defOps, mockFn),
      dBtn = getByTestId('drop-btn'),
      dCtn = getByTestId('drop-contain'),
      dOBtn = getByTestId(defOps[0].val);
    addOffsetParent(dBtn);
    fireEvent.click(dBtn);
    expect(dCtn.style.display).toBeTruthy();
    fireEvent.click(dOBtn);
    expect(mockFn).toHaveReturnedWith(defOps[0].val);
    expect(mockFn).toHaveBeenCalledTimes(1);
  });

});

