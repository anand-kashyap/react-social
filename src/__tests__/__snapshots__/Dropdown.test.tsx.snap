// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Component should take a snapshot 1`] = `
<DocumentFragment>
  <button
    class="dropdown-trigger btn-flat more-details"
    data-target="dropdowntest1"
    data-testid="drop-btn"
  >
    <i
      class="material-icons-outlined"
    >
      more_horiz
    </i>
  </button>
  <ul
    class="dropdown-content"
    data-testid="drop-contain"
    id="dropdowntest1"
    tabindex="0"
  >
    <li
      tabindex="0"
    >
      <button
        class="drop-item"
        data-testid="one"
      >
        <i
          class="material-icons-outlined"
        >
          all_out
        </i>
         
        <span>
          one
        </span>
      </button>
    </li>
    <li
      tabindex="0"
    >
      <button
        class="drop-item"
        data-testid="delete"
      >
        <i
          class="material-icons-outlined"
        >
          delete
        </i>
         
        <span>
          delete
        </span>
      </button>
    </li>
  </ul>
</DocumentFragment>
`;
